### ISSUE TITLE
MSM with random errors in Tomcat 7.0.54

### ISSUE THREAD

**Issue by federicoleon** [magro/memcached-session-manager#279]
Hi there! I have MSM running on 5 Tomcat servers, (with Tomcat version **7.0.54**) and they all have the same MSM configuration in context.xml:

`<Manager className="de.javakaffee.web.msm.MemcachedBackupSessionManager"
memcachedNodes="n1:11.21.234.5:2987,n2:11.21.234.7:2386"
 sticky="false"
lockingMode="auto"
requestUriIgnorePattern=".*\.(png|gif|jpg|css|js)$"
sessionBackupAsync="false"
sessionBackupTimeout="500"
operationTimeout="5000" />`

I'm using the following jars:
**memcached-session-manager-1.8.3.jar**
**memcached-session-manager-tc7-1.8.3.jar**
**spymemcached-2.11.1.jar**

Most of the sessions are stored and retrieved correctly but I also can see many errors like these:

`WARNING: Could not store session DEA55F86635D2F9AF3FCA48A96F4A943-n1 in memcached.
Note that this session was relocated to this node because the original node was not available.`

`net.spy.memcached.internal.CheckedOperationTimeoutException: Timed out waiting for operation - failing node: 11.21.234.5:2987`

`INFO: Session needs to be relocated as node 3381 is not available, loading backup session for 24DA131521FCA74BB658076BCA6AC195-n7`

`WARNING: Could not find session in local session map.`

`WARNING: Could not store session 35C868845A97F44BE102B901BA7EFFBC-n1 in memcached.
net.spy.memcached.internal.CheckedOperationTimeoutException: Timed out waiting for operation - failing node: 11.21.234.5:2987`

`WARNING: Could not store session 3D4852A7028EABF85CFB67F6CDC917C3-n1 in memcached.
net.spy.memcached.internal.CheckedOperationTimeoutException: Timed out waiting for operation - failing node: 11.21.234.5:2987`

`INFO: Session needs to be relocated as node n7 is not available, loading backup session for 855E7055718C749619242F276F9D3D80-n7`

`INFO: Session needs to be relocated as node n8 is not available, loading backup session for 049A1E98DD047CA75388572B47229CDC-19032480`

`WARNING: Could not find session in local session map.`

`INFO: Session needs to be relocated as node n8 is not available, loading backup session for 81F4C2170DC3CBD4C652D0610060E4B1-n8`

`WARNING: Could not find session in local session map.`

`WARNING: Could not find session in local session map.`

I know for sure that my memcached nodes are working correctly since I have tools monitoring these memcached.

Any suggestion will be MUCH appreciated.

**Comment by magro** [comment]
There are nodes mentioned that are not configured: 3381, n7, n8. Have you changed your msm config, or what might be the reason for this? Perhaps this is related to the issue.

If spymemcached throws a CheckedOperationTimeoutException, we cannot do anything about this. So the question is what's causing the timeouts.

**Comment by federicoleon** [comment]
Sorry, my bad. I've updated my project with following jars:
`asm-3.2.jar`

`kryo-1.04.jar`

`kryo-serializers-0.11.jar`

`memcached-session-manager-1.8.3.jar`

`memcached-session-manager-tc7-1.8.3.jar`

`minlog-1.2.jar`

`msm-kryo-serializer-1.8.3.jar`

`reflectasm-1.01.jar`

`spymemcached-2.11.1.jar`

And this configuration:

`<Manager className="de.javakaffee.web.msm.MemcachedBackupSessionManager"
    memcachedNodes="n1:11.21.234.5:2987,n2:11.21.234.7:2386"
    sticky="false"
    lockingMode="auto"
    requestUriIgnorePattern=".*\.(png|gif|jpg|css|js)$"
    sessionBackupAsync="false"
    sessionBackupTimeout="500"
    operationTimeout="5000"
    transcoderFactoryClass="de.javakaffee.web.msm.serializer.kryo.KryoTranscoderFactory"/>`

Now I see these errors:

`WARNING: Could not deserialize session with id DD121F36ED6626F95960D197890B29CB-n1 from memcached, session will be purged from storage.`

`WARNING: Could not deserialize session with id 59CA8FCE598A637261DBD6CA5F556EA4-n2 from memcached, session will be purged from storage.`

Any ideas? Thanks in advance!

**Comment by magro** [comment]
This happens when the previously serialized session cannot be deserialized, e.g. because classes have changed. Might this be the case here?

**Comment by federicoleon** [comment]
@magro thank you so much for your quick reply!
I don't think this is the case. When I runned the deploy with the change described on my previous comment I saw these errors. With that change I've changed all my classes to be serializable and I've changed the serialVersionUID in all of them. In theory, after this change every file should be serialized with the new serialVersionUID, right? So every new session should be using the new classes.
I can't get how I'm getting these errors if, in theory, the sessions are generated with the new logic and classes.
Is there any way to debug or see a more descriptive log?
Thanks and sorry for bothering you.

**Comment by federicoleon** [comment]
I've enabled full logs for MSM and this is what I see after the error messages I've commented above:

`WARNING: Could not deserialize session with id 59CA8FCE598A637261DBD6CA5F556EA4-n2 from memcached, session will be purged from storage.`

`de.javakaffee.web.msm.TranscoderDeserializationException: com.esotericsoftware.kryo.SerializationException: Unable to deserialize object of type: java.util.concurrent.ConcurrentHashMap
at de.javakaffee.web.msm.serializer.kryo.KryoTranscoder.deserializeAttributes(KryoTranscoder.java:291)`

`Caused by: com.esotericsoftware.kryo.SerializationException: Unable to deserialize object of type: java.util.concurrent.ConcurrentHashMap
at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:593)`

`Caused by: com.esotericsoftware.kryo.SerializationException: Encountered unregistered class ID: 114
at com.esotericsoftware.kryo.Kryo.readClass(Kryo.java:463)`

**Comment by magro** [comment]
Re implementing Serializable and serialVersionUID: this just doesn't matter when using kryo. You might switch to Java serialization if kryo issues are too hard to solve. 

If you'd like to continue to try to get kryo running I'd suggest that you switch to the latest version 1.9.0-RC1 (uses the latest kryo version instead of kryo1) and see if the situation is the same. In this case I can better help then. Btw, you must also flush memcached after the upgrade.

**Comment by federicoleon** [comment]
Thanks for your reply and for your support with this issue. I've moved back to Java serialization. Let me give you some context, I'm a litle bit lost with how this tool works.

I have 3 servers running on different machines. I have my context.xml defined as docs says:

`<Manager className="de.javakaffee.web.msm.MemcachedBackupSessionManager"
       memcachedNodes="n1:xxx.xxx.xxx:1234,n2:xxx.xxx.xxx:5678"
       sticky="false"
       lockingMode="none"
       requestUriIgnorePattern=".*\.(png|gif|jpg|css|js)$"
       sessionBackupAsync="false"
       sessionBackupTimeout="500"
       operationTimeout="5000"/>`

1. My application has a single form where I request for some information. When I post this form I ask for the session, (stored in memcached, if there is no session MSM creates one and retrieves it). I store all data in this session. **`JSESSIONID:"B22CF3DDE4A127D1E135123CBA4CD83A-n2"`**
2. After this I go to another part of my application where I display a confirmation and if user confirms I ask for the session, **`JSESSIONID:"B22CF3DDE4A127D1E135123CBA4CD83A-n2"`** (with data already completed in the previous step), and post this.

Right now I can see my sessionID B22CF3DDE4A127D1E135123CBA4CD83A-n2 stored on second memcached. When I try to post I retrieve the session from memcached but it has lost all the information that I saved in the previous step. These steps may occur in different Tomcat servers.

I think that the session is being deleted from memcached between step 1 and step 2, so when I try to retrieve it in step 2 I'm actually creating it again. ¿Is this possible? ¿How long they are stored? ¿Is there a way to set for how long I want the sessions to be stored?

Also I have lockingMode="auto" and I haven't configured a prefix for sessions. ¿Is this correct?

Thank you SO much for your help.
